systemverilog-plugin/UhdmAst.cc

```diff
@@ -4540,6 +4594,83 @@ void UhdmAst::process_unsupported_stmt(const UHDM::BaseClass *object, bool is_er
+    std::string renamed_enum = "";
+
+    visit_one_to_one({vpiTypespec}, obj_h, [&](AST::AstNode *node) {
+        if (!node)
+            return;
+
+        if (node->type == AST::AST_WIRE && node->str == "") {
+            // skip AST_WIREs that don't have a name,
+            // this happens when processing default values of parameters
+            delete node;
+            return;
+        }
+
+        if (node->type == AST::AST_ENUM) {
+            // Enum typedefs are composed of AST_ENUM and AST_TYPEDEF where the enum shall be renamed,
+            // so that the original name used in code is assigned to the AST_TYPEDEF node,
+            // and a mangled name is assigned to the AST_ENUM node.
+            renamed_enum = node->str + "$enum";
+        }
+
+        current_node->children.push_back(node->clone());
+
+        // The child stores information about the type assigned to the parameter
+        //   this information will be used to rename the module
+
+        // find the typedef for `node` in the upper scope and copy it to .children of the AST_PARAMETER node
+        // if unable to find the typedef, continue without error as this could be a globally available type
+
+        // vpiParent object is the module which has the typespec,
+        vpiHandle module_def_h = vpi_handle(vpiParent, obj_h);
+        // the typedef is stored in the outer module
+        vpiHandle parent_parent_h = vpi_handle(vpiParent, module_def_h);
+        vpi_release_handle(module_def_h);
+
+        std::string parent_parent_name = get_name(parent_parent_h, true);
+        vpi_release_handle(parent_parent_h);
+        if (!shared.top_nodes.count(parent_parent_name)) {
+            // no outer scope available, stop searching for the typedef
+            return;
+        }
+
+        for (auto child : shared.top_nodes[parent_parent_name]->children) {
+            // name of the type we're looking for
+            if (child->str == node->str && child->type == AST::AST_TYPEDEF) {
+                current_node->children.push_back(child->clone());
+                break;
+            }
+        }
+        delete node;
+    });
+
+    if (renamed_enum != "") {
+        for (auto child : current_node->children) {
+            if (child->type == AST::AST_TYPEDEF) {
+                child->children[0]->attributes[ID::enum_type] = AST::AstNode::mkconst_str(renamed_enum);
+            }
+            if (child->type == AST::AST_ENUM) {
+                child->str = renamed_enum;
+            }
+        }
+    }
+    if (current_node->children.size() == 0) {
+        // nothing to do with the parameter, the actual value was substituted everywhere already in uhdm
+        if (current_node->attributes.count(attr_id::is_type_parameter)) {
+            current_node->attributes.erase(attr_id::is_type_parameter);
+        }
+        delete current_node;
+        current_node = nullptr;
+    }
```

Comment:
I think we should error out when we can't find parameter type, otherwise we could wrongly rename module.
```suggestion
    if (current_node->children.size() == 0) {
        log_file_error(current_node->filename, current_node->location.first_line, "Empty parameter type!");
    }
```

Reply:
I've introduced changes to support anonymous types, and this is no longer relevant.
Parameters with anonymous types do have 0 children but now this is allowed, as the `is_type_parameter` attribute is used and that's enough.